core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationServiceImpl.java

```diff
@@ -0,0 +1,159 @@
+ */
+package org.sonarsource.sonarlint.core.sync;
+
+import com.google.common.eventbus.Subscribe;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.sonarsource.sonarlint.core.ServerApiProvider;
+import org.sonarsource.sonarlint.core.branch.SonarProjectBranchServiceImpl;
+import org.sonarsource.sonarlint.core.clientapi.SonarLintClient;
+import org.sonarsource.sonarlint.core.clientapi.client.sync.DidSynchronizeConfigurationScopeParams;
+import org.sonarsource.sonarlint.core.commons.Language;
+import org.sonarsource.sonarlint.core.event.ActiveSonarProjectBranchChanged;
+import org.sonarsource.sonarlint.core.progress.ProgressNotifier;
+import org.sonarsource.sonarlint.core.progress.TaskManager;
+import org.sonarsource.sonarlint.core.repository.config.ConfigurationRepository;
+import org.sonarsource.sonarlint.core.serverapi.ServerApi;
+import org.sonarsource.sonarlint.core.serverconnection.ServerConnection;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+public class SynchronizationServiceImpl {
+
+  private final SonarLintClient client;
+  private final ConfigurationRepository configurationRepository;
+  private final SonarProjectBranchServiceImpl branchService;
+  private final ServerApiProvider serverApiProvider;
+  private final TaskManager taskManager;
+  private ScheduledExecutorService scheduledSynchronizer;
+  private Path storageRoot;
+  private Path workDir;
+  private Set<Language> enabledLanguagesInConnectedMode;
+  private boolean taintVulnerabilitiesEnabled;
+  private Map<String, ServerConnection> connectionsById = new HashMap<>();
+
+  public SynchronizationServiceImpl(SonarLintClient client, ConfigurationRepository configurationRepository, SonarProjectBranchServiceImpl branchService,
+    ServerApiProvider serverApiProvider) {
+    this.client = client;
+    this.configurationRepository = configurationRepository;
+    this.branchService = branchService;
+    this.serverApiProvider = serverApiProvider;
+    this.taskManager = new TaskManager(client);
+  }
+
+  public void initialize(Path storageRoot, Path workDir, Set<Language> enabledLanguagesInConnectedMode, boolean taintVulnerabilitiesEnabled) {
+    this.storageRoot = storageRoot;
+    this.workDir = workDir;
+    this.enabledLanguagesInConnectedMode = enabledLanguagesInConnectedMode;
+    this.taintVulnerabilitiesEnabled = taintVulnerabilitiesEnabled;
+    scheduledSynchronizer = Executors.newSingleThreadScheduledExecutor(r -> new Thread(r, "Sonar Project Synchronizer"));
```

Comment:
The thread name is a bit limiting IMO. What about "SonarLint Local Storage Synchronizer"?